pennylane/compiler/qjit_api.py

```diff
@@ -111,3 +112,173 @@ def circuit(theta):
     compilers = AvailableCompilers.names_entrypoints
     qjit_loader = compilers[compiler]["qjit"].load()
     return qjit_loader(fn=fn, *args, **kwargs)
+
+
+def while_loop(*args, **kwargs):
+    """A :func:`~.qjit` compatible while-loop for PennyLane programs.
+
+    .. note::
+
+        This function only supports the Catalyst compiler. See
+        :func:`catalyst.while_loop` for more details.
+
+        Please see the Catalyst :doc:`quickstart guide <catalyst:dev/quick_start>`,
+        as well as the :doc:`sharp bits and debugging tips <catalyst:dev/sharp_bits>`
+        page for an overview of the differences between Catalyst and PennyLane.
+
+    This decorator provides a functional version of the traditional while
+    loop, similar to ``jax.lax.while_loop``. That is, any variables that are
+    modified across iterations need to be provided as inputs and outputs to
+    the loop body function:
+
+    - Input arguments contain the value of a variable at the start of an
+      iteration
+
+    - Output arguments contain the value at the end of the iteration. The
+      outputs are then fed back as inputs to the next iteration.
+
+    The final iteration values are also returned from the
+    transformed function.
+
+    The semantics of ``while_loop`` are given by the following Python pseudo-code:
+
+    .. code-block:: python
+
+        def while_loop(cond_fun, body_fun, *args):
+            while cond_fun(*args):
+                args = body_fn(*args)
+            return args
+
+    Args:
+        cond_fn (Callable): the condition function in the while loop
```

Comment:
Do we need `body_fn` too?

Reply:
This is an interesting question and I'm curious how we handle it in PennyLane.

_Technically_, this object is a function that only takes 1 parameter - `cond_fn` - and returns another function. This new function is what accepts the `body_fn` as a parameter. 

Of course this applies to all (parametrized) decorators. What is the best way then?

- `Args: <decorator parameters>`, `returns: function`
- `Args: <decorator parameters> + <function to decorate>`, `returns: function `
- `Args: <decorator parameters> + <function to decorate>`, `returns: <decorated function results>`

Reply:
I'm not sure I have a good answer :laughing: So given that how it is currently is the technically correct description, we can leave this as is.